specs/sharding/beacon-chain.md

```diff
@@ -0,0 +1,690 @@
+
+### `AttestationData`
+
+```python
+class AttestationData(Container):
+    slot: Slot
+    index: CommitteeIndex
+    # LMD GHOST vote
+    beacon_block_root: Root
+    # FFG vote
+    source: Checkpoint
+    target: Checkpoint
+    # Shard header root
+    shard_header_root: Root
+```
+
+### `BeaconBlockBody`
+
+```python
+class BeaconBlockBody(merge.BeaconBlockBody):  # [extends The Merge block body]
+    shard_headers: List[SignedShardHeader, MAX_SHARDS * MAX_SHARD_HEADERS_PER_SHARD]
+```
+
+### `BeaconState`
+
+```python
+class BeaconState(merge.BeaconState):  # [extends The Merge block body]
+    # [Updated fields]
+    previous_epoch_attestations: List[PendingAttestation, MAX_ATTESTATIONS * SLOTS_PER_EPOCH]
+    current_epoch_attestations: List[PendingAttestation, MAX_ATTESTATIONS * SLOTS_PER_EPOCH]
+    # [New fields]
+    previous_epoch_pending_shard_headers: List[PendingShardHeader, MAX_SHARDS * MAX_SHARD_HEADERS_PER_SHARD * SLOTS_PER_EPOCH]
+    current_epoch_pending_shard_headers: List[PendingShardHeader, MAX_SHARDS * MAX_SHARD_HEADERS_PER_SHARD * SLOTS_PER_EPOCH]
+    grandparent_epoch_confirmed_commitments: Vector[Vector[DataCommitment, SLOTS_PER_EPOCH], MAX_SHARDS]
+    shard_gasprice: uint64
+    current_epoch_start_shard: Shard
+```
+
+## New containers
+
+The shard data itself is network-layer only, and can be found in the [P2P specification](./p2p-interface.md).
+The beacon chain registers just the commitments of the shard data.
+
+### `DataCommitment`
+
+```python
+class DataCommitment(Container):
+    # KZG10 commitment to the data
+    point: BLSCommitment
+    # Length of the data in samples
+    length: uint64
+```
+
+### `ShardHeader`
+
+```python
+class ShardHeader(Container):
+    # Slot and shard that this header is intended for
+    slot: Slot
+    shard: Shard
+    # The actual data commitment
+    commitment: DataCommitment
+    # Proof that the degree < commitment.length
+    degree_proof: BLSCommitment
+```
+
+TODO: add shard-proposer-index to shard headers, similar to optimization done with beacon-blocks.
+
+### `SignedShardHeader`
+
+```python
+class SignedShardHeader(Container):
+    message: ShardHeader
+    signature: BLSSignature
+```
+
+### `PendingShardHeader`
+
+```python
+class PendingShardHeader(Container):
+    # Slot and shard that this header is intended for
+    slot: Slot
+    shard: Shard
+    # KZG10 commitment to the data
+    commitment: DataCommitment
+    # hash_tree_root of the ShardHeader (stored so that attestations can be checked against it)
+    root: Root
+    # Who voted for the header
+    votes: Bitlist[MAX_VALIDATORS_PER_COMMITTEE]
+    # Has this header been confirmed?
+    confirmed: bool
+```
+
+## Helper functions
+
+### Misc
+
+#### `next_power_of_two`
+
+```python
+def next_power_of_two(x):
+    return 2 ** ((x - 1).bit_length())
+```
+
+#### `compute_previous_slot`
+
+```python
+def compute_previous_slot(slot: Slot) -> Slot:
+    if slot > 0:
```

Comment:
Should `0` be replaced with `GENESIS_SLOT` in this function?

Reply:
It is `Slot(0)` below, and for the underflow problem more so than configuration (`GENESIS_SLOT` is pretty much a constant at this point)